### PR TITLE
feat(cdp): add microsoft teams template

### PR DESCRIPTION
**Pull request by MarconLP**
## Changes

- adds microsoft teams hog function

## How did you test this code?

added test

### CODE REVIEW

File: posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from posthog.cdp.templates.hog_function_template import HogFunctionTemplate, HogFunctionSubTemplate, SUB_TEMPLATE_COMMON
+
+template: HogFunctionTemplate = HogFunctionTemplate(
+    status="beta",
```
Comment (mariusandra) [inline posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py:4]: Same question here as for discord. Perhaps we want to make this free?

**Review by mariusandra — APPROVED** [review]

File: posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py (hunk 1 above)
Comment (MarconLP) [reply posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py:4]: Yeah, we should make this free.

I don't seem to find a way to check if a string starts with `https://%.webhook.office.com/webhookb2/` in hog.

I tried this code: 
```
if (not inputs.webhookUrl like 'https://%.webhook.office.com/webhookb2/') {
    throw Error('Invalid url');
}
```

But that will allow this URL `https://webhook.site/571d559e-238d-4305-90c3-d36ec91fe7d5#https://max.webhook.office.com/webhookb2/`

Comment (mariusandra) [reply posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py]: You'll need the `%` in the end as well... so `url like 'https://%.webhook.office.com/webhookb2/%'`

Comment (MarconLP) [reply posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py]: I tried the syntax from the standup today but it's still possible to trick the check by including the correct URL as a fragment:
2024-10-18 at 12 03 29@2x

It doesn't seem like it's possible to filter for `string startsWith 'https://%.webhook.office.com/webhookb2/%'`

Comment (mariusandra) [reply posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py]: Right... in this case the first `%` in `https://%.` is effectively matches the string `'webhook.site/57...d5#https://max'`. We'll need better URL parsing to get past this. 

Some ideas:
- Use a regex and be strict about what characters are allowed
- Split the initial URL and discard everything after `?` and `#` symbols
- ClickHouse has a bunch of [URL functions](https://clickhouse.com/docs/en/sql-reference/functions/url-functions) and we could port some of them over.

Comment (MarconLP) [reply posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py]: I am going to use this condition:
```
if (not match(inputs.webhookUrl, '^https://[^/]+.webhook.office.com/webhookb2/.*'))
```

File: posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+template: HogFunctionTemplate = HogFunctionTemplate(
+    status="free",
+    id="template-microsoft-teams",
+    name="Microsoft Teams",
+    description="Sends a message to a Microsoft Teams channel",
+    icon_url="/static/services/microsoft-teams.png",
+    category=["Customer Success"],
+    hog="""
+if (not match(inputs.webhookUrl, '^https://[^/]+.webhook.office.com/webhookb2/.*')) {
+    throw Error('Invalid url');
```
Comment (mariusandra) [inline posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py:12]: We'll save a few support pingpongs if in the error we also say what's the right format.

**Review by mariusandra — APPROVED** [review]

File: posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py (hunk 2 above)
Comment (MarconLP) [reply posthog/cdp/templates/microsoft_teams/template_microsoft_teams.py]: added the correct format into the error message

**Comment by MarconLP** [comment]
Seems like Microsoft is actually deprecating webhooks in favor of workflows. Although they delayed the deprecation several times https://devblogs.microsoft.com/microsoft365dev/retirement-of-office-365-connectors-within-microsoft-teams/

I've modified the template to use the workflow endpoint instead.

**Pull request merged**